src/hastings_vacuum.erl

```diff
@@ -93,38 +114,79 @@ code_change(_OldVsn, St, _Extra) ->
 
 
 handle_db_event(DbName, created, _St) ->
-    gen_server:cast(?MODULE, {cleanup, DbName}),
+    gen_server:cast(?MODULE, {cleanup, DbName, []}),
     {ok, nil};
 handle_db_event(DbName, deleted, _St) ->
-    gen_server:cast(?MODULE, {cleanup, DbName}),
+    gen_server:cast(?MODULE, {cleanup, DbName, [{context, delete}]}),
     {ok, nil};
 handle_db_event(_DbName, _Event, _St) ->
     {ok, nil}.
 
 
-maybe_start_cleaner(#st{cleaner=undefined}=St) ->
-    case queue:is_empty(St#st.queue) of
+maybe_start_selective_cleaner(#st{selective_cleaner=undefined}=St) ->
+    case queue:is_empty(St#st.selective_clean) of
         false ->
-            start_cleaner(St);
+            start_selective_cleaner(St);
         true ->
             {noreply, St}
     end;
 
-maybe_start_cleaner(St) ->
+maybe_start_selective_cleaner(St) ->
     {noreply, St}.
 
 
-start_cleaner(St) ->
-    {{value, DbName}, NewQ} = queue:out(St#st.queue),
-    {Pid, _} = spawn_monitor(?MODULE, clean_db, [DbName]),
-    {noreply, St#st{queue = NewQ, cleaner = Pid}}.
+maybe_start_full_cleaner(#st{full_cleaner=undefined}=St) ->
+    case queue:is_empty(St#st.full_clean) of
+        false ->
+            start_full_cleaner(St);
+        true ->
+            {noreply, St}
+    end;
+
+maybe_start_full_cleaner(St) ->
+    {noreply, St}.
+
+
+start_selective_cleaner(St) ->
+    {{value, DbName}, NewQ} = queue:out(St#st.selective_clean),
+    {Pid, _} = spawn_monitor(?MODULE, clean_db, [DbName, []]),
+    {noreply, St#st{selective_clean = NewQ, selective_cleaner = Pid}}.
+
+
+start_full_cleaner(St) ->
+    {{value, ShardDbName}, NewQ} = queue:out(St#st.full_clean),
+    {Pid, _} = spawn_monitor(
+        ?MODULE, clean_sharddb, [ShardDbName, [{context, delete}]]
+    ),
+    {noreply, St#st{full_clean = NewQ, full_cleaner = Pid}}.
 
 
 clean_db(DbName) ->
+    clean_db(DbName, []).
+
+clean_db(DbName, _Options) ->
+    delete_inactive_indexes(DbName, get_active_sigs(DbName)).
+
+
+clean_sharddb(ShardDbName) ->
+    clean_db(ShardDbName, []).
+
+clean_sharddb(ShardDbName, Options) ->
+    Context = couch_util:get_value(context, Options, compaction),
+    DoRecovery = config:get_boolean("couchdb",
+        "enable_database_recovery", false),
+    case {Context, DoRecovery} of
```

Comment:
You're gonna have a case match issue here since you don't have a clause matching the compaction for Context.